src/utils/handleSubmit.js

```diff
@@ -0,0 +1,43 @@
+import Popup from "../components/Popup.js";
+
+export function handleSubmit(
+  submitAction,
+  popupInstance, // This should be an instance of Popup or PopupDeleteCard
+  submitText = "Save",
+  loadingText = "Saving..."
+) {
+  // Check if popupInstance is an instance of Popup
+  if (!(popupInstance instanceof Popup)) {
+    console.error("The provided popupInstance is not an instance of Popup.");
+    return Promise.reject("Invalid popup instance");
+  }
+
+  const popup = popupInstance._popupElement;
+  const submitButton = popup.querySelector(".modal__save-button");
+
+  if (!submitButton) {
+    console.error("Submit button not found in popup.");
+    return Promise.reject("Submit button not found");
+  }
+
+  submitButton.disabled = true;
+  submitButton.textContent = loadingText;
+
+  // Ensure submitAction returns a promise
+  return Promise.resolve(submitAction())
+    .then(() => {
+      popupInstance.close();
+      submitButton.disabled = false;
+      submitButton.textContent = submitText;
+
+      const form = popup.querySelector("form");
+      if (form) form.reset();
+    })
+    .catch((err) => {
+      console.error(`Error: ${err}`);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```